Documentation (README.md) sufficient to understand and reproduce the
analysis conducted to generate one figure from your project
- Folder structure
At a high level, what is in each folder and subfolder?
- Installation
How do I run your code?
What software and package versions do I need to install?

## 20.440 Final Project Repository - Phillip Clauer & Viraat Goel
### Overview
This repository contains the code, data, figures, and other files relevant to Phillip C. & Viraat G.'s 20.440 final project for the
Spring 2020 semester. The project, entitled "*Determining sex-specific differentially expressed transcripts in ALS patients*", aims
to analyze ALS transcriptomics to identify how differential gene expression between the male and female sexes may be contributing to
the greater observed prevalence of ALS in men.

The data folder contains raw gene expression data from two different databases
([GSE124439](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124439) and
[GSE122649](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122649)), the code folder contains Python 3 scripts to analyze
the data in the databases, and the figures folder contains script-generated plots to visually capture data analysis.

### Data
The data utilized in this project is derived from two databases, both of which contain gene expression data generated using RNA-seq
of post-mortem cortexes of ALS patients and individuals without neurological disorders:
- [GSE124439](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124439)
- [GSE122649](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122649)
Compiled by [Tam et al. (2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6866666/), the data were generated by the New York
Genome Center (GSE124439) and the Cold Spring Harbor Lab (GSE122649). Specifically, RNA was extracted from cortical tissue and
paired-end sequencing was performed using an Illumina HiSeq-2500 to quantify transcripts as a readout for gene expression. Accordingly,
the data present in the databases consist of individual .txt files for each patient sample with each file containing 28952 gene names
and the corresponding number of raw transcript reads for that gene.
