# 20.440 Final Project Repository - Phillip Clauer & Viraat Goel
## Overview
This repository contains the code, data, figures, and other files relevant to Phillip C. & Viraat G.'s 20.440 final project for the Spring 2020 semester. The project, entitled "*Determining sex-specific differentially expressed transcripts in ALS patients*", aims to analyze ALS transcriptomics to identify how differential gene expression between the male and female sexes may be contributing to the greater observed prevalence of ALS in men.

The data folder contains raw gene expression data from two different databases ([GSE124439 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124439) and [GSE122649](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122649)), the code folder contains Python 3 scripts to analyze the data in the databases, and the figures folder contains script-generated plots to visually capture data analysis.

## Data
The data utilized in this project is derived from two databases, both of which contain gene expression data generated using RNA-seq of post-mortem cortexes of ALS patients and individuals without neurological disorders:
- [GSE124439](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124439)
- [GSE122649](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122649)

Compiled by [Tam et al. (2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6866666/), the data were generated by the New York Genome Center (GSE124439) and the Cold Spring Harbor Lab (GSE122649). Specifically, RNA was extracted from cortical tissue and paired-end sequencing was performed using an Illumina HiSeq-2500 to quantify transcripts as a readout for gene expression. Accordingly, the data present in the databases consist of individual .txt files for each patient sample with each file containing 28952 gene names and the corresponding number of raw transcript reads for that gene.

**Citation**: Tam OH, Rozhkov NV, Shaw R, Kim D, et al. Postmortem Cortex Samples Identify Distinct Molecular Subtypes of ALS: Retrotransposon Activation, Oxidative Stress, and Activated Glia. *Cell Rep* 2019 Oct 29; 29(5):1164-1177.e5. PMID:
[31665631](https://www.ncbi.nlm.nih.gov/pubmed/31665631)

## Folder Structure
This repository contains three main folders - data, code, and figures.
- **Data**: As described above, our data is pulled from two databases. As a result, the *data* folder contains two subfolders, each with the raw data from one of the two databases. The GSE124439_RAW subfolder contains data from 176 patient samples while the GSE122649_RAW subfolder contains data from 38 patient samples, with all sample data being contained in individual .txt files.
- **Code**: The code folder contains Python 3 scripts to analyze the data in the data folder. So far, the only script in the folder is one which generates a log10-scaled heatmap of all the read counts in either database; this serves as our proof-of-principle to demonstrate that we are able to access, load, and visualize the data.
- **Figures**: The figures folder contains all plots, graphs, and figures that are generated by the script(s) in the code folder as visualization and analysis aids for the data in the data folder. Currently, the folder contains .png images of the full-dataset heatmaps of each dataset.

## Installation
The *als_Data_Heatmap_2pager.py* script runs using Python 3 and is dependent upon standard Python packages including pandas, os, math, numpy, matplotlib, and seaborn. Provided that the data, code, and figures folders are available as currently structured in the repository (e.g. by cloning this repository locally and running it in a virtual environment), the script should execute without issues. A MakeFile and requirements.txt will be added for ease of execution as this project grows in size and complexity. Please contact Viraat Goel (viraat@mit.edu) with any questions or if you have any trouble running the script. Thank you!
